src/transformers/models/modernbert/modular_modernbert.py

```diff
@@ -0,0 +1,1233 @@
+        labels: Optional[torch.Tensor] = None,
+        ignore_index: int = -100,
+    ):
+        return _pad_modernbert_output(
+            inputs=inputs, indices=indices, batch=batch_size, seqlen=seqlen, labels=labels, ignore_index=ignore_index
+        )
+
+
+class ModernBertModel(ModernBertPreTrainedModel):
+    def __init__(self, config: ModernBertConfig):
+        super().__init__(config)
+        self.config = config
+        self.embeddings = ModernBertEmbeddings(config)
+        self.layers = nn.ModuleList(
+            [ModernBertEncoderLayer(config, layer_id) for layer_id in range(config.num_hidden_layers)]
+        )
+        self.final_norm = nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+        self.gradient_checkpointing = False
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.embeddings.tok_embeddings
+
+    def set_input_embeddings(self, value):
+        self.embeddings.tok_embeddings = value
+
+    def _init_weights(self, module: Optional[nn.Module] = None, reset_params: Optional[bool] = None):
+        if module and hasattr(module, "_init_weights"):
+            super()._init_weights(module, reset_params)
+        elif isinstance(reset_params, bool):
+            self.embeddings._init_weights(reset_params=reset_params)
+
+            if reset_params:
+                self.final_norm.reset_parameters()
+
+    def forward(
+        self,
+        input_ids: torch.LongTensor = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        return_dict: Optional[bool] = None,
+        indices: Optional[torch.Tensor] = None,
+        cu_seqlens: Optional[torch.Tensor] = None,
+        max_seqlen: Optional[int] = None,
+        batch_size: Optional[int] = None,
+        seq_len: Optional[int] = None,
+    ) -> Union[Tuple[torch.Tensor, ...], BaseModelOutput]:
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if batch_size is None and seq_len is None:
+            batch_size, seq_len = input_ids.shape[:2]
+
+        if attention_mask is None:
+            attention_mask = torch.ones((batch_size, seq_len), device=input_ids.device, dtype=torch.bool)
+
+        repad = False
+        if self.config.unpad_inputs:
+            if indices is None and cu_seqlens is None and max_seqlen is None:
+                repad = True
+                if self.config.unpad_no_grad:
+                    input_ids, indices, cu_seqlens, max_seqlen, *_ = self._unpad_inputs_no_grad(
+                        input_ids, attention_mask
+                    )
+                else:
+                    input_ids, indices, cu_seqlens, max_seqlen, *_ = self._unpad_inputs(input_ids, attention_mask)
+        elif position_ids is None:
+            position_ids = torch.arange(seq_len, device=input_ids.device)
+
+        hidden_states = self.embeddings(input_ids)
+
```

Comment:
We don't have the eager attention yet, but it probably requires expanding the attention mask to 4D somewhere around here, e.g.:
```python
        # expand attention_mask
        if self.config._attn_implementation == "eager" and attention_mask is not None:
            # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
            attention_mask = _prepare_4d_attention_mask(attention_mask, hidden_states.dtype)
```

I didn't get the other attentions working yet @warner-benjamin, perhaps you can try and tackle those.

Reply:
Added via a2051d6b4a0c771ecd36ba05603fe5823110acb5